dev/src/bundle.ts

```diff
@@ -0,0 +1,225 @@
+ */
+export class BundleBuilder {
+  // DocumentReference and Query added to the bundle.
+  private docRefs: DocumentReference[] = [];
+  private queries: Map<string, Query> = new Map<string, Query>();
+
+  // DocumentSnapshot and QuerySnapshot added to the bundle.
+  private docSnaps: DocumentSnapshot[] = [];
+  private querySnaps: Map<string, QuerySnapshot> = new Map<
+    string,
+    QuerySnapshot
+  >();
+
+  // Resulting documents for the bundle.
+  private documents: Map<string, BundledDocument> = new Map();
+  // Named queries saved in the bundle.
+  private namedQueries: Map<string, firestore.INamedQuery> = new Map();
+
+  // The latest read time among all bundled documents and queries.
+  private latestReadTime: Timestamp = Timestamp.fromProto({
+    seconds: 0,
+    nanos: 0,
+  });
+
+  constructor(private id: string) {}
+
+  add(doc: DocumentReference): BundleBuilder;
+  add(docSnap: DocumentSnapshot): BundleBuilder;
+  add(queryName: string, querySnap: QuerySnapshot): BundleBuilder;
+  add(queryName: string, query: Query): BundleBuilder;
+  add(
+    docOrName: DocumentReference | DocumentSnapshot | string,
+    query?: QuerySnapshot | Query
+  ) {
+    if (docOrName instanceof DocumentReference) {
+      this.docRefs.push(docOrName);
+    } else if (docOrName instanceof DocumentSnapshot) {
+      this.docSnaps.push(docOrName);
+    } else if (query instanceof Query) {
+      this.queries.set(docOrName, query);
+    } else if (query instanceof QuerySnapshot) {
+      this.querySnaps.set(docOrName, query);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private pushLengthPrefixedString(readable: Readable, payload: string) {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+
+    readable.push(Buffer.concat([lengthBuffer, buffer]));
+  }
+
+  stream(): NodeJS.ReadableStream {
```

Comment:
If you rebase this on top of the Node 10 branch you can use `async iterators` which are a user friendly version of  streams. While we didn't discuss this during the design, do you want to explore what this option would look like?

https://javascript.info/async-iterators-generators

Example: https://github.com/googleapis/nodejs-firestore/blob/node10/dev/src/v1/firestore_client.ts#L1335

Reply:
As discussed, we will use this stream impl for now. Async iterators are nice, but bunldes are usually already serialized when user have to deal with them, so make it part of interface does not bring a lot benefit usually come in other cases.

Reply:
Sounds good. FWIW, Async Iterators can easily consume Streams and customers can build their own Async Iterator API.